src/main/java/org/primeframework/mvc/security/cookiesession/UserIDCookieSessionSecurityContext.java

```diff
@@ -0,0 +1,236 @@
+      return CookieExtendResult.Keep;
+    }
+    var halfWayPoint = signInTime.plusMinutes(sessionTimeout.toMinutes() / 2);
+    return now.isAfter(halfWayPoint) ? CookieExtendResult.Extend : CookieExtendResult.Keep;
+  }
+
+  @Override
+  public Object getCurrentUser() {
+    return Optional.ofNullable(getSessionContainer())
+                   .map(c -> c.user)
+                   .orElse(null);
+  }
+
+  private InMemorySessionContainer getSessionContainer() {
+    var container = (InMemorySessionContainer) this.request.getAttribute(UserKey);
+    if (container != null) {
+      return container;
+    }
+    var cookie = this.sessionCookie.get(this.request);
+    if (cookie == null) {
+      return null;
+    }
+    try {
+      var deserializedContainer = CookieTools.fromJSONCookie(cookie,
+                                                             SerializedSessionContainer.class,
+                                                             true,
+                                                             encryptor,
+                                                             objectMapper);
+      var shouldExtend = shouldExtendCookie(ZonedDateTime.now(this.clock),
+                                            deserializedContainer.signInInstant,
+                                            sessionMaxAge,
+                                            sessionTimeout);
+      switch (shouldExtend) {
+        case Extend:
+          // same cookie value, but with longer time (set on cookie proxy in constructor)
+          this.sessionCookie.add(request, response, cookie);
+          break;
+        case Invalid:
+          // past max age, don't trust it at all
+          deleteCookies();
+          return null;
+      }
+      var user = retrieveUserById(deserializedContainer.userId);
+      // in our same request cycle, we avoid a DB call by holding on to this
+      container = new InMemorySessionContainer(deserializedContainer, user);
+      this.request.setAttribute(UserKey, container);
+      return container;
+    } catch (BadPaddingException e) {
+      // if the cookie encryption key changes (DB change, etc.) then we need new cookies, otherwise we cannot
+      // decrypt and a user will be stuck trying to get back in
+      this.deleteCookies();
+      return null;
+    } catch (Exception e) {
+      throw new ErrorException(e);
+    }
+  }
+
+  /**
+   * Hydrates/retrieves the user based
+   *
+   * @param id unique ID for the user
+   * @return the object representing the user
+   */
+  protected abstract Object retrieveUserById(UUID id);
+
+  @Override
+  public String getSessionId() {
+    return Optional.ofNullable(getSessionContainer()).map(c -> c.sessionId).orElse(null);
+  }
+
+  @Override
+  public boolean isLoggedIn() {
+    // user causes a DB query
+    return getSessionId() != null;
+  }
+
+  @Override
+  public void login(Object context) {
+    try {
+      if (context instanceof IdentifiableUser user) {
```

Comment:
Personal preference. But if we require a specific type, I would do that and the immediately throw an exception instead of doing it in the `else` block.

Reply:
Done